src/Substance.hs

```diff
@@ -540,17 +600,24 @@ subSeparate = foldr separate ([], [])
 
 
 -- | 'parseSubstance' runs the actual parser function: 'substanceParser', taking in a program String, parses it, semantically checks it, and eventually invoke Alloy if needed. It outputs a collection of Substance objects at the end.
-parseSubstance :: String -> String -> VarEnv -> IO (SubObjects, VarEnv)
+parseSubstance :: String -> String -> VarEnv -> IO (SubObjects, (VarEnv, SubEnv))
 parseSubstance subFile subIn varEnv =
                case runParser substanceParser subFile subIn of
                Left err -> error (parseErrorPretty err)
                Right xs -> do
+                   divLine
                    putStrLn "Substance AST: \n"
                    mapM_ print xs
+                   let subTypeEnv = check xs varEnv
+                       c          = loadObjects xs subTypeEnv
+                       subSubEnv   = loadSubEnv xs
```

Comment:
maybe rename `subSubEnv` to `subDynEnv`? it behaves a little more like a dynamic/evaluation environment (even though it's not exactly that). `subEqEnv` (for equality) could also work